File: src/main/kotlin/com/linkit/commons/utils/RoleSetConverter.java
package com.linkit.commons.utils;

import com.linkit.domain.user.model.Role;

import javax.persistence.AttributeConverter;
import java.util.Arrays;
import java.util.HashSet;
import java.util.Set;
import java.util.stream.Collectors;

public class RoleSetConverter implements AttributeConverter<Set<Role>, String> {
    private static final String SPLIT_CHAR = ",";

    @Override
    public String convertToDatabaseColumn(Set<Role> attribute) {
        return attribute.stream().map(Enum::name).collect(Collectors.joining(SPLIT_CHAR));
    }

    @Override
    public Set<Role> convertToEntityAttribute(String string) {
        if (null == string) {
            return new HashSet<>();
        }
        return Arrays.stream(string.split(SPLIT_CHAR)).map(Role::valueOf).collect(Collectors.toSet());
    }
}

File: src/main/kotlin/com/linkit/domain/user/model/RoleSetConverter.kt
package com.linkit.domain.user.model

import javax.persistence.AttributeConverter

class RoleSetConverter : AttributeConverter<Set<Role>, String> {

    override fun convertToDatabaseColumn(attribute: Set<Role>): String {
        return attribute.joinToString(SPLIT_CHAR) { it.name }
    }

    override fun convertToEntityAttribute(string: String): Set<Role> {
        return string.split(",").map { Role.valueOf(it) }.toSet()
    }

    companion object {
        private const val SPLIT_CHAR = ","
    }
}